src/components/Pagination/Pagination.js

```diff
@@ -0,0 +1,133 @@
+import Link from 'next/link'
+import { useRouter } from 'next/router'
+import PropTypes from 'prop-types'
+import { useState } from 'react'
+
+import makePaginatedRoute from '~/util/router/makePaginatedRoute'
+
+import styles from './Pagination.module.scss'
+
+const DEFAULT_PAGE_SIZE = 25
+
+function Pagination (props) {
+  const {
+    author,
+    category,
+    page,
+    pageInput,
+    pageSize,
+    onUpdatePageSize,
+    route,
+    totalPages = 1,
+  } = props
+
+  const [currentPage, setCurrentPage] = useState(page)
+
+  const router = useRouter()
+
+  const showPrevPage = (page > 1 ? '' : 'hidden')
+  const showNextPage = (page < totalPages ? '' : 'hidden')
+
+  let prevPage = makePaginatedRoute({
+    route, author, category, page: Math.max(1, page - 1),
+  })
+
+  let nextPage = makePaginatedRoute({
+    route, author, category, page: Math.min(page + 1, totalPages),
+  })
+
+  if (pageSize !== DEFAULT_PAGE_SIZE) {
+    prevPage = makePaginatedRoute({
+      route, author, category, page: Math.max(1, page - 1), limit: pageSize,
+    })
+
+    nextPage = makePaginatedRoute({
+      route, author, category, page: Math.min(page + 1, totalPages), limit: pageSize,
+    })
+  }
```

Comment:
Instead of accepting `route`, `author`, and `category` as props, I think it would be better to pass the control of route formation off to the parent. Instead I'd recommend a couple of method props similar to:

```tsx
<Pagination
  onMakePrevRoute={(params) => makeRoute('leaderboard', params)}
  onMakeNextRoute={(params) => makeRoute('leaderboard', params)} />
```

This has a few benefits. Not only does it simplify the props surface, it also lets the pagination component conform to the route definition of the page, rather than requiring a strict page structure